ethash/src/cache.rs

```diff
@@ -185,6 +185,7 @@ fn make_memmapped_cache(path: &Path, num_nodes: usize, ident: &H256) -> io::Resu
 
 	let mut memmap = unsafe { MmapMut::map_mut(&file)? };
 
+	#[allow(clippy::cast_ptr_alignment)] // TODO: Why is this okay?
```

Comment:
We should use [ptr::read_unaligned](https://doc.rust-lang.org/std/ptr/fn.read_unaligned.html) instead

Reply:
I'm still a bit confused about what's going on here. But it really looks wrong to me:
* [`initialize_memory`](https://github.com/paritytech/parity-ethereum/blob/33bfcec4ab3fe0ad4deb69a038edae998d0536a0/ethash/src/cache.rs#L303) internally creates a `slice::from_raw_parts_mut` of type `&mut [Node]`.
* `MmapMut` does not guarantee alignment, otherwise it wouldn't have an [`align_to_mut`](https://docs.rs/memmap/0.7.0/memmap/struct.MmapMut.html#method.align_to_mut) method.
* [`from_raw_parts_mut`](https://doc.rust-lang.org/std/slice/fn.from_raw_parts_mut.html) explicitly requires the data to be aligned.

`read_unaligned` would copy the data: My impression is that most of this code was written as an optimization to avoid copying. If we are fine with copying, we could probably do away with most of the unsafe code here?

Reply:
I thought mmap is page-aligned http://man7.org/linux/man-pages/man2/mmap.2.html, but maybe @Vurich could provide more input.

> MmapMut does not guarantee alignment, otherwise it wouldn't have an align_to_mut method.

This is a method on a slice though (I guess it shows up in the docs, because `MapMut` implements `Deref` to a slice), which can be arbitrary aligned.

Reply:
And I agree we should as much unsafe code as possible, we have some benchmarks, see e.g. #10861. But let's do it in a separate PR.

Reply:
Ah, [I see](https://docs.rs/memmap/0.7.0/memmap/struct.MmapMut.html#deref-methods)! Sorry! :man_facepalming: 
Then I guess I should just add a comment pointing to the mmap man page? (But what about platforms other than Linux?)

Reply:
> I thought mmap is page-aligned http://man7.org/linux/man-pages/man2/mmap.2.html, but maybe @Vurich could provide more input.
> 
> > MmapMut does not guarantee alignment, otherwise it wouldn't have an align_to_mut method.
> 
> This is a method on a slice though (I guess it shows up in the docs, because `MapMut` implements `Deref` to a slice), which can be arbitrary aligned.

The link you want is https://pubs.opengroup.org/onlinepubs/9699919799/functions/mmap.html - while on Linux it is defined to be page-aligned, mmap is a POSIX syscall and so on non-Linux POSIX environments it might fail. I'd add an alignment check (with error on fail) here, LLVM has alignment knowledge for libc functions for all architectures it supports so we can rely on it to remove the alignment check where appropriate.